### DISCUSSION TITLE
USB controller?

### DISCUSSION THREAD

**Discussion by horsiezoo** [nk64/GT4286#7] [Q&A]
Has anyone tried plugging in a USB controller? I tried 3 , Xbox one, two different Logitechs, the most I can get is the ui freezes. I noticed you can power the system via usb (I was checking if can get a serial connection) sorry if this has already been covered :-)

**Comment by nk64** [comment]
I have only tried plugging in a keyboard with no luck.

**Comment by Hello-dot-Jpg** [comment]
I think my old old PC controller sort of works, I can glitch the menu if press down on it while pressing down on the p1 stick where it'll continue to scroll the games list even after the p1 stick is released.

I can't seem to find any 4 player games, even ones that I am pretty sure were four player are displayed as only 2 player when run on the emulator (Simpsons and DND)

**Reply by horsiezoo** [reply]
Yes that’s what happened to me, self scrolling until unplug usb, kind of good to know that I haven’t damaged the usb port through experimentation. Even tried a powered usb hub , no difference . Very oddly designed device!

**Comment by cgb** [comment]
I wondered if this unit can act as controller input to a PC?  Has anyone tried?

i.e. https://youtu.be/YxQ2QyLYFiU?si=zpDA7tNhvHvNX_dM&t=1071

**Reply by horsiezoo** [reply]
Yes already tried, USB A male to male cable - USB from console to pc USB port, nothing detected, but noted console can be powered from its USB port, eg : unplug power cable while usb cable connected to pc, except then the power switch does nothing. Theres something strange about the USB on this, I assume the arcade buttons/stick use USB unless its some wierd hack. This is proving to be quite fun just figuring out how this thing works.

**Reply by Hello-dot-Jpg** [reply]
I think it would work if you connected to the daughter i/o board directly, but as you've said before it looks like the device can be powered by USB so you'd have to disconnect the ribbon cable so that the main board doesn't also get powered when you connect it to your PC

Or you could get a USB KVM switcher or something like that, you'd just have to add a usb male connector to the daughter board and female to the main board

if you wanted to get fancy you could use a 4 pole selector and a screw in usb socket to switch between PC mode and Internal hardware mode

img/hardware/io-board-front.jpg

**Reply by cgb** [reply]
Somebody noted at:
https://www.ozbargain.com.au/comment/16111194/redir

indicating they've successfully done this.  It doesn't make a lot of sense to me (not super electronically minded) but maybe someone here understand what they wrote.  Would love to know more about that process as I'm thinkig this unit will be most useful to me if it oculd be peripherals to a more capable x86 emulator PC or RPI.

**Reply by horsiezoo** [reply]
Very cool thx ," [ThighHighDemise](https://www.ozbargain.com.au/user/81412) on 29/12/2024 - 09:43
For those more interested in potentially picking one of these up to adapt into a usb controller to connect to an external PC (or raspberry pi/whatever else) instead of dealing with the included SOC, it's quite easy to accomplish. All the USB lines are clearly labelled on the IO board. Hooked up to a Linux/Windows PC shows up as two separate HID devices (p1/p2) and generic drivers work fine. For $50, I'm happy."  
Alright time to get my soldering iron out, I'm guessing they mean simply soldering a usb connector as you say or solder a cut off usb cable to the board, I'll give this a go later

**Reply by Hello-dot-Jpg** [reply]
When I go pick up my second one, I might make the modifications, if I do, I'll make a guide on how to do it.

**Comment by horsiezoo** [comment]
[ls.txt](https://github.com/user-attachments/files/18383810/ls.txt)
Ok got a usb controller working, I found a ps2 controller to usb adapter I had lying around and… well it works. I’m guessing it will work with controllers that don’t need any special drivers or have any extra functionality. 
Edit : ps2 and psx controllers work with a usb adapter , something like this

https://www.aliexpress.com/item/32995987125.html?spm=a2g0o.productlist.main.13.7e7bZ9qeZ9qeM6&algo_pvid=deaaebef-8a27-4d42-a0f2-93aa7f27c65e&algo_exp_id=deaaebef-8a27-4d42-a0f2-93aa7f27c65e-6&pdp_npi=4%40dis%21NZD%214.03%213.06%21%21%212.21%211.68%21%402101ea7117365562043165986e06e7%2110000006040127424%21sea%21NZ%210%21ABX&curPageLogUid=4wCigyLjMeB2&utparam-url=scene%3Asearch%7Cquery_from%3A

there’s also a two input version, I might try ordering one to see if it can recognise two controllers on one usb port 

https://www.aliexpress.com/item/1005007218526531.html?spm=a2g0o.productlist.main.11.7e7bZ9qeZ9qeM6&algo_pvid=deaaebef-8a27-4d42-a0f2-93aa7f27c65e&algo_exp_id=deaaebef-8a27-4d42-a0f2-93aa7f27c65e-5&pdp_npi=4%40dis%21NZD%217.44%217.44%21%21%2130.02%2130.02%21%402101ea7117365562043165986e06e7%2112000039846446044%21sea%21NZ%210%21ABX&curPageLogUid=im2rkHc9OOlz&utparam-url=scene%3Asearch%7Cquery_from%3A

![IMG_2075](https://github.com/user-attachments/assets/0bd03710-e82d-4fb3-a6e4-384525da5280)


Connecting two controllers via ps2 multitap does not make console see two controllers 

![IMG_2076](https://github.com/user-attachments/assets/f265dbd0-1442-4ed1-b8a3-bbfd9bc3bf1e)

edit : ok this is very cool, having a controller plugged in, press “1p coin” button on console, in pcsx emulator changes frameskip off/1/auto 
Start button: take screenshot 
R button : turns off fps display 

gambatte , fbneo unfortunately doesn’t seem to have this functionality 

Having a usb controller plugged in means only 1 player, console arcade controls either disabled or remapped. 

ls .sh dump with controller plugged in shows what's changed :

/dev/bus:
drwxr-xr-x    5 0        0              100 Jan  1 00:00 usb/

/dev/bus/usb:
drwxr-xr-x    2 0        0               80 Jan  1 00:00 001/
drwxr-xr-x    2 0        0               80 Jan  1 00:00 002/
drwxr-xr-x    2 0        0               60 Jan  1 00:00 003/

/dev/bus/usb/001:
crw-------    1 0        0         189,   0 Jan  1 00:00 001
crw-------    1 0        0         189,   1 Jan  1 00:00 002

/dev/bus/usb/002:
crw-------    1 0        0         189, 128 Jan  1 00:00 001
crw-------    1 0        0         189, 129 Jan  1 00:00 002

/dev/bus/usb/003:
crw-------    1 0        0         189, 256 Jan  1 00:00 001

/dev/input:
crw-------    1 0        0          13,  64 Jan  1 00:00 event0
crw-------    1 0        0          13,  65 Jan  1 00:00 event1
crw-------    1 0        0          13,  66 Jan  1 00:00 event69
crw-------    1 0        0          13,  67 Jan  1 00:00 event70
crw-------    1 0        0          13,  68 Jan  1 00:00 event71
crw-------    1 0        0          13,  69 Jan  1 00:00 event72

**Reply by horsiezoo** [reply]
I think I've worked out why some controllers work and others don't :
lsusb with a device plugged in to usb port 2:
Bus 003 Device 001: ID 1d6b:0002
Bus 002 Device 002: ID 0810:0001
Bus 001 Device 001: ID 1d6b:0002
Bus 001 Device 002: ID 071b:3203
Bus 002 Device 001: ID 1d6b:0002

The basic usb on this doesn't seem to read the descriptor on the device e.g the mp3 player i had plugged in shows up on raspberry pi lsusb as "domain technologies, inc. rockchip " on the retro arcade : just the hex id. 

(grain of salt happy to be proven wrong)

**Reply by nk64** [reply]
I think that usb descriptors are just not included in the busybox version of lsusb to save on size

https://unix.stackexchange.com/questions/220759/lsusb-doesnt-list-device-names

**Comment by nk64** [comment]
I just worked out that my USB port on my GT4286 is dead so that is why I wasn't able the replicate the effect of plugging in a USB keyboard and pressing ALT+CTRL+DEL.

Luckily I have two more GT4286s.

**Reply by Hello-dot-Jpg** [reply]
If you wanted to keep using that one for testing you could probably cut the end off a USB hub and solder that on to the main board then add a tail for the controllers and connect that to the hub

**Reply by nk64** [reply]
I'll have a go at repairing it (one day), but until I have a go I don't know if I've fried a chip pin or if it is just a dodgy connection to the board.

On the one hand, I think it has been dead for a long time (I tried plugging in a keyboard over two weeks ago) but on the other hand yesterday I experienced what seemed to be a short circuit which caused the console to reboot as I was plugging in a USB3 Dual ethernet dongle so I thought that that could have broken it.

Since that short circuit I've been a bit gunshy and only plugging USB devices in when it is turned off at the power point because I don't think the machine is electrically isolated very well.

**Reply by horsiezoo** [reply]
That’s kind of interesting, possibly usb device drawing more power than available. power source is 5v 1amp so after powering board etc probably doesn’t leave the full usb rated 500ma. 
Now i wish i had one of those usb current output test devices out of pure curiosity. 
 Do you have a multimeter at hand? If so is the connector outputting 5v. if no with console powered off measure resistance between the 5v vcc pin and a gnd pin on board, should be high. 
 Perhaps there’s a protection fuse on the usb circuit which could explain only one port being affected as opposed to both.
Going forward a possible safe guard could be a powered usb hub to lessen the strain on the power supply 

—edit : This console doesn’t like usb hubs, for one a powered hub “back powers” console - supplies power to its usb port , and console appears to assign a different “event number” to a controller plugged in to said hub therefore doesn’t “work” for very much want of a better word.

**Comment by nk64** [comment]
I have uploaded an `evtest` utility that I have customised to dump the capabilities of all devices and a User Script (previously referred to as Custom Scripts) to run it.

You can read a bit about it [here](https://github.com/nk64/GT4286/blob/main/Input%20Devices.md)

You can also take the meat of the [user script](https://github.com/nk64/GT4286/blob/main/sdcard_tweaks/roms/MAME/test-evtest-dump-capabilities.sh):
```
evtest="${GT4286UTIL_HOME}/bin/evtest"
"${evtest}" --allcaps > "${output_dir}/test-evtest-dump-capabilities.txt" 2>&1
```
and stick it at the beginning of `/gt4286util/scripts/gmenu2x_wrapper.sh` to see how plugging in a joystick or keyboard affects the numbering and capabilities of the device.

**Reply by horsiezoo** [reply]
[output.zip](https://github.com/user-attachments/files/18406475/output.zip)
[outputps2.zip](https://github.com/user-attachments/files/18406622/outputps2.zip)

Cool tool thx, so definitely detected correctly (I zipped the output logs -(output.zip=logitech, outputps2.zip=ps2 adaptor)and ran the detection test, literally mashed all buttons whole time screen cycling colours (good workaround there) ...nothing.. which tracks as this control pad doesn't seem to be supported. I'll run it with PS2 to usb thing and see what's different.
 you are correct Busybox lsusb only shows device id  

usb control pad :
Available devices:
  /dev/input/event0:	mcu_ts
  /dev/input/event1:	mcu_ts
  /dev/input/event8:	Logitech Logitech Dual Action
  /dev/input/event69:	Twin USB Gamepad
  /dev/input/event70:	Twin USB Gamepad



PS2 to usb detected as 
Available devices:
  /dev/input/event0:	mcu_ts
  /dev/input/event1:	mcu_ts
  /dev/input/event69:	Twin USB Joystick
  /dev/input/event70:	Twin USB Joystick
  /dev/input/event71:	Twin USB Gamepad
  /dev/input/event72:	Twin USB Gamepad

Which means ..... drivers ? Can't really confirm until have tried a few different controllers

**Reply by nk64** [reply]
Good work.
I think that in `output2.zip/test-evtest-event69.txt` you managed to capture some output of one of the controllers on the PS2 to usb hub.

You will need to adjust the capture script to account for the extra/changed event numbering when you add different devices.
It is not smart enough yet to capture all the available /dev/input/event* devices only the ones out of the box (0,1,69,70). so you have to add extra lines for 71 and 72.

Also, I think that it is good idea to be methodical in the joystick movements and buttons you press so that you can pick out the pattern in the captured output and compare it to the values in `/joystick.cfg` (first using the inbuilt controller) then your own.

Also you might like to run `debug-logcat` and see what it says about `get mastrd fd`
In my example it looks like it is scanning up to event70 and has picked event69 as `mastrt df` whatever that means, I think that maybe it means that controller 69 will control the gmenu2x  (but not sure).
```
    # D/zkgui   (  936): Open key:/dev/input/event0 
    # D/zkgui   (  936): open [/dev/input/event0] [12] sucess
    # D/zkgui   (  936): get mastrt fd [/dev/input/event0]
    # D/zkgui   (  936): Open key:/dev/input/event1 
    # D/zkgui   (  936): open [/dev/input/event1] [13] sucess
    # D/zkgui   (  936): Open key:/dev/input/event6 
    # D/zkgui   (  936): Open key:/dev/input/event8 
    # D/zkgui   (  936): Open key:/dev/input/event67 
    # D/zkgui   (  936): Open key:/dev/input/event68 
    # D/zkgui   (  936): Open key:/dev/input/event69 
    # D/zkgui   (  936): open [/dev/input/event69] [14] sucess
    # D/zkgui   (  936): get mastrt fd [/dev/input/event69]
    # D/zkgui   (  936): Open key:/dev/input/event70 
    # D/zkgui   (  936): open [/dev/input/event70] [15] sucess
```

Unfortunately I don't have any PS1 controllers to help out (though I do have a Dual PS1 to USB adapter for some reason :shrug:)

**Reply by horsiezoo** [reply]
Very interesting thanks! Just looking in to that. Yes it’s amazing the random stuff we accumulate that eventually becomes useful. Fortunately PlayStation controllers aren’t exactly rare 2nd hand 

Ok  getting somewhere ! Added the seemingly device specific event numbers (namely 8 in this case- Logitech) and output :  
[test-evtest-event8.txt](https://github.com/user-attachments/files/18417101/test-evtest-event8.txt)

[test-evtest-dump-capabilities.txt](https://github.com/user-attachments/files/18417103/test-evtest-dump-capabilities.txt)

Will look further into this a bit later.
Thanks for the assist once again, really appreciate your hard work, basically if it wasn't for you this console would likely be in my pile of "ill considered purchasing decisions"

**Reply by nk64** [reply]
now it can be your gateway drug into reverse engineering an embedded linux device.

**Comment by xbekoolx** [comment]
My unit boots and joysticks or any buttons does not works. I have ps5 and i tried connecting ps5 controller into this. It does work but laggy. Not smooth in menu.

Can someone help how can i check if my board itself is damaged? I plugged 40 pin connector in pandora box and it works perfectly there, but in this board. I even solder two usb1 connections but does not work either. I am so fed up with this.

**Reply by nk64** [reply]
Hi @xbekoolx,
I wonder where you got with this? 
If you are still interested in troubleshooting (after so long), please take a look at [this](38) discussion thread.
Cheers,
NK64.
